print-edsa.c

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright (c) 1988, 1989, 1990, 1991, 1992, 1993, 1994, 1995, 1996, 1997
+ *	The Regents of the University of California.  All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that: (1) source code distributions
+ * retain the above copyright notice and this paragraph in its entirety, (2)
+ * distributions including binary code include the above copyright notice and
+ * this paragraph in its entirety in the documentation or other materials
+ * provided with the distribution, and (3) all advertising materials mentioning
+ * features or use of this software display the following acknowledgement:
+ * ``This product includes software developed by the University of California,
+ * Lawrence Berkeley Laboratory and its contributors.'' Neither the name of
+ * the University nor the names of its contributors may be used to endorse
+ * or promote products derived from this software without specific prior
+ * written permission.
+ * THIS SOFTWARE IS PROVIDED ``AS IS'' AND WITHOUT ANY EXPRESS OR IMPLIED
+ * WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#define NETDISSECT_REWORKED
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <tcpdump-stdinc.h>
+
+#include <string.h>
+
+#include "interface.h"
+#include "addrtoname.h"
+#include "ether.h"
+#include "ethertype.h"
+#include "extract.h"			/* must come after interface.h */
+
+static const char tstr[] = "[|DSA]";
+
+/*
+ * Extended Distributed Switch Archiecture.
+ *
+ * A Marvell propriatary header used for passing packets to/from
+ * specific ports of a switch.
+ *
+ */
+struct  edsa_pkthdr {
+	u_char bytes[6];
+
+	/* Bytes 0 and 1 are reserved and should contain 0 */
+#define TAG(edsa) 	(edsa->bytes[2] >> 6)
+#define TAG_TO_CPU	0
+#define TAG_FROM_CPU	1
+#define TAG_FORWARD	3
+#define SRC_TAG(edsa) 	((edsa->bytes[2] >> 5) & 0x01)
+#define SRC_DEV(edsa)	(edsa->bytes[2] & 0x1f)
+#define SRC_PORT(edsa)	((edsa->bytes[3] >> 3) & 0x01f)
+#define TRUNK(edsa)	((edsa->bytes[3] >> 2) & 0x01)
+#define CODE(edsa)	((edsa->bytes[3] & 0x06) |	\
+			 ((edsa->bytes[4] >> 4) & 0x01))
+#define CODE_BDPU	0
+#define CODE_IGMP_MLD	2
+#define CODE_ARP_MIRROR	4
+#define CFI(edsa)	(edsa->bytes[3] & 0x01)
+#define PRI(edsa)	(edsa->bytes[4] >> 5)
+#define VID(edsa)	(((u_short)(edsa->bytes[4] & 0xf) << 8 |						edsa->bytes[5]))
+	u_short ether_type;
+};
```

Comment:
You might of figured this out, but i copied print-arp.c as a starting point. It mixes its structure definition and #defines like this. But i can move them out.